xml/issue4264.xml

```diff
@@ -38,6 +38,131 @@ an implementation to perform such an optimization. As a consequence, it is accep
 owning wrapper from a non-owning one, where implementing such an optimization may not be possible.
 </p>
+
+<superseded>
+<p>
+This wording is relative to <paper num="N5008"/>.
+</p>
+<ol>
+
+<li><p>Modify <sref ref="[func.wrap.general]"/> as indicated:</p>
+
+<blockquote>
+<p>
+-2- Let <tt>t</tt> be an object of a type that is a specialization of <tt>function</tt>,
+<tt>copyable_function</tt>, <del>or</del> <tt>move_only_function</tt><ins>, or <tt>function_ref</tt></ins>,
+such that the target object <tt>x</tt> of <tt>t</tt> has a type that is a specialization of <tt>function</tt>,
+<tt>copyable_function</tt>, <del>or</del><tt>move_only_function</tt><ins>, or <tt>function_ref</tt></ins>.
+Each argument of the invocation of <tt>x</tt> evaluated as part of the invocation of <tt>t</tt>
+may alias an argument in the same position in the invocation of <tt>t</tt> that has the same type,
+even if the corresponding parameter is not of reference type.
+</p>
+</blockquote>
+
+</li>
+</ol>
+
+</superseded>
+
+<note>2024-05-21; Tomasz's comment and upates proposed resolution</note>
+
+<p>
+After implementing double indirecation avoidance in the libstdc++, I have realized
+that above wording change is insufficient to cover all user observable effects of
+the change. Revelant quiote from the <a href="https://gcc.gnu.org/pipermail/libstdc++/2025-May/061561.html">
+Avoid double indirection in function_ref</a>  from libsdc++ mailing lists:
+</p>
+
+<blockquote>
+<p>
+To avoidance of double indirection requires that constructed <tt>function_ref</tt>,
+refers directly to the target function of the source, instead of source,
+and this is visible after the assigment:
+</p>
+
+<pre>
+void foo() noexcept;
+void bar() noexcept;
+
+std::function_ref&lt;void() noexcept&gt; sr(&amp;foo);
+std::function_ref&lt;void()&gt; dr(sr);
+dr(); // calls `foo` regardless of implementation
+
+sr = &amp;bar;
+sr(); // calls `bar`
+dr(); // still calls `foo` if we avoid indirection,
+      // calls `bar` if we do not
+</pre>
+
+<p>
+Similary for <tt>move_only_function</tt>/<tt>copyable_function</tt> source:
+</p>
+
+<pre>
+std::move_only_function&lt;void()&gt; sm;
+std::function_ref&lt;void()&gt; dm(sm);
+
+dm(); // UB because `sm` is empty
+
+sm = &amp;foo;
+
+dm(); // remains UB if we avoid indirection,
+      // calls `bar` if we do not.
+</pre>
+
+<p>
+While we may want to allow skipping indirection for function_ref,
+as this produces same behavior as in case for copy constructor (matching
+signatures):
+</p>
+
+<pre>
+void foo() noexcept;
+void bar() noexcept;
+
+std::function_ref&lt;void() noexcept&gt; sr(&amp;foo);
+std::function_ref&lt;void() noexcept&gt; dr(sr); // copy-cosntructor
+dr(); // calls `foo` regardless of implementation
+
+sr = &amp;bar;
+sr(); // calls `bar`
+dr(); // still calls `foo` if we avoid indirection
+</pre>
+
+
+<p>
+I do not think this is acceptable for <tt>move_only_function</tt>. So most
+likely we will and only first commit.
```

Comment:
I'm not sure what the end of this sentence means, is there a word missing?

Reply:
As this is a quote from my email, I do not want it to be fixed in place. As this is not relevant, I replaced it with `&hellip;`.